Lots of rewriting for the Scala 3 Reference (#15993)
lampepfl/dotty: docs/_docs/reference/contextual/derivation.md

We say that `Tree` is the _deriving type_ and that the `Eq`, `Ordering` and `Show` instances are _derived instances_.

## Exact mechanism
More formally, for a class/trait/object/enum `DerivingType[T_1, ..., T_N] derives TC` (if `DerivingType` does not take parameters, we define `N = 0`), a derived instance is created in `DerivingType`'s companion object (or `DerivingType` itself if it is an object)

What the derived instance looks like depends on the specifics of `DerivingType` and `TC`, first the arity of `TC`:

### `TC` takes 1 parameter

Therefore `TC` is defined as `TC[F[A_1, ..., A_K]]` (`TC[F]` if `K == 0`), there are two further cases depending on the `A_i`s:

#### `F` and all arguments of `DerivingType` have kind `*`

The generated instance is then:
```scala
given [T_1: TC, ..., T_N: TC]: TC[DerivingType[T_1, ..., T_N]] = TC.derived
```

If `N == 0`, we understand the above to mean:
```scala
given TC[DerivingType] = TC.derived
```

<!-- #### You can pair arguments of `F` and `DerivingType` starting from the right such that they have the same kinds pairwise, and all arguments of at least one of them are used up -->

> #### `F` and `DerivingType` have their `m` rightmost arguments which have the same kind pairwise and `N >= 0`, `K > 0`, and `m == N` and/or `m == K`

Reviewer: ```suggestion
#### `F` and `DerivingType`'s `m` rightmost arguments are the same kind pairwise and `N >= 0`, `K > 0`, and `m == N` and/or `m == K`
```

The generated instance is then:
If `m == N == K`:
```scala
given TC[DerivingType] = TC.derived
```
If `m == N < K`:
```scala
given TC[[A_1, ..., A_K] =>> DerivingType[A_(K-N+1), ..., A_K]] = TC.derived
```

> TODO: Speak about `N == 0` being allowed, since then `(K-N+1) > K`

Reviewer: fix the todo :)

We say that `Tree` is the _deriving type_ and that the `Eq`, `Ordering` and `Show` instances are _derived instances_.

> More formally, when a class/trait/object/enum `DerivingType` `derives` a type class `TC`, the following given instance is created in `DerivingType`'s companion object (or `DerivingType` itself if it is an object):

Reviewer: ```suggestion
More formally, for a class/trait/object/enum `DerivingType derives TC`, the following given instance is created in `DerivingType`'s companion object (or `DerivingType` itself if it is an object):
```

* if `DerivingType` doesn't have type parameters
```scala
given TC[DerivingType] = TC.derived
```

> * if `DerivingType` has type parameters `[T_1, ... T_N]`

Reviewer: These rules are further refined in https://github.com/lampepfl/dotty/blob/main/compiler/src/dotty/tools/dotc/typer/Deriving.scala

Reviewer: mostly for higher kinded types

```scala
given [T_1: TC, ... T_N: TC]: TC[DerivingType[T_1, ... T_N]] = TC.derived
```

> `TC.derived` is then in charge of creating an instance of `TC[DerivingType]`, potentially using term and/or type inference.

Reviewer: ```suggestion
`TC.derived` should be a term that conforms to the expected type `TC[DerivingType]`, potentially using term and/or type inference.
```

**Note:** `TC.derived` is a normal method call, therefore if there are multiple definitions of `TC.derived`, overloading resolution applies.

> **Note:** `TC.derived` can be used manually, for example to implement `Ordering` for `Option`:

Reviewer: this is discouraged pls

```scala

> given [T: Ordering]: Ordering[Option[T]] = Ordering.derived

Reviewer: please noo

Reviewer: I realise this was moved around, but I suppose it should be strongly recommended to use `derives` clause as much as possible

given [T: Ordering]: Ordering[Option[T]] = Ordering.derived
```

All data types can have a `derives` clause. This document focuses primarily on data types which also have a given instance
of the `Mirror` type class available.
### `Mirror`

`Mirror` type class instances provide information at the type level about the components and labelling of the type.

`Mirror` instances provide information at the type level about the components and labelling of the type.
They also provide minimal term level infrastructure to allow higher level libraries to provide comprehensive
derivation support.

> Instances of the `Mirror` type class are generated automatically by the compiler

Reviewer: mirror is a type class, I'm not sure why you removed all the times it is said

+ The methods `ordinal` and `fromProduct` are defined in terms of `MirroredMonoType` which is the type of kind-`*`
  which is obtained from `MirroredType` by wildcarding its type parameters.

> ### implementing `derived` with `Mirror`

Reviewer: ```suggestion
### Implementing `derived` with `Mirror`
```
